Improve streamable

richardkiss opened the pull request.

richardkiss committed:
Avoid importing `test_constants` as it takes a long time.

richardkiss committed:
Factor out `parse_*` functions.

richardkiss committed:
First crack at refactoring `Streamable.parse`.

richardkiss committed:
Don't add `_parse_functions` attribute to `Streamable`.

This no longer requires an extra `_parse_functions` attribute on a
`Streamable`, as it may be confusing serializers or other functions
that use `__annotations__`.

richardkiss committed:
Fix lint problems with `black`.

richardkiss committed:
Fix `parse_tuple`.

richardkiss committed:
Defer some parsing failures to parse time rather than class-creation time.

richardkiss committed:
Tidy up & remove some obsolete stuff.

richardkiss committed:
Decorate `RequestBlocks` as `streamable`.

mariano54 commented on chia/util/streamable.py:
```diff
@@ -152,80 +155,131 @@ def streamable(cls: Any):
+    return t
+
+
+def parse_bool(f):
```
All these functions need type annotations

arvidn replied:
https://github.com/Chia-Network/chia-blockchain/pull/3222

rostislav committed:
Fix wrong uses of Streamable class

Revert an earlier commit and error out on class creation in case a
Streamable subclass is instantiated incorrectly, e.g. containing a
non-serializable member.

Fix cases where Streamable parent class was forgotten.

rostislav committed:
Fix wrong types when creating DerivationRecord and WalletCoinRecord

aqk commented on chia/util/streamable.py:
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
+
+def parse_bool(f):
+    bool_byte = f.read(1)
+    assert bool_byte is not None and len(bool_byte) == 1  # Checks for EOF
```
Will these asserts activate in any run mode, or is it like C, where they can be deactivated in some run modes?

richardkiss replied:
`assert` is removed in `python -OO` mode, which we don't do. So we shouldn't use `assert` for flow control. However, this problem was here before this change, so it doesn't make these semantics any worse.

aqk commented on chia/util/streamable.py:
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
-            bytes_read = f.read(list_size)
-            assert bytes_read is not None and len(bytes_read) == list_size
-            return bytes_read
+            return parse_bytes
```
This section is nice.

aqk commented on chia/wallet/wallet_coin_store.py:
```diff
@@ -126,17 +127,23 @@ async def set_spent(self, coin_name: bytes32, height: uint32):
+    def coin_record_from_row(self, row: sqlite3.Row) -> WalletCoinRecord:
+        coin = Coin(bytes32(bytes.fromhex(row[6])), bytes32(bytes.fromhex(row[5])), uint64.from_bytes(row[7]))
+        return WalletCoinRecord(
+            coin, uint32(row[1]), uint32(row[2]), bool(row[3]), bool(row[4]), WalletType(row[8]), row[9]
```
What is the type of row[9] - I notice it is the only one not explicitly cast

arvidn replied:
a comment with the schema of this table would be very helpful, but I suppose this is not quite in scope for this PR

aqk commented on chia/wallet/wallet_puzzle_store.py:
```diff
@@ -120,11 +120,11 @@ async def get_derivation_record(self, index: uint32, wallet_id: uint32) -> Optio
                 G1Element.from_bytes(bytes.fromhex(row[1])),
-                row[3],
-                row[4],
+                WalletType(row[3]),
```
WalletType: nice.

aqk commented on chia/util/streamable.py:
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
+        fields = {}
+
+    for _, f_type in fields.items():
+        parse_functions.append(cls.function_to_parse_one_item(f_type))
```
Is the intent here to automatically add parse functions for only the primitive types that are contained as members?

richardkiss replied:
The general idea is to traverse the `__annotations__` tree just once at class instantiation time, and put into a list the sub-parsing functions which do parsing and not introspection.

aqk reviewed (dismissed):
Looks good. One thing I don't understand in the first question.

arvidn commented on chia/protocols/timelord_protocol.py:
```diff
@@ -74,7 +74,7 @@ class NewEndOfSubSlotVDF(Streamable):
 @dataclass(frozen=True)
 @streamable
-class RequestCompactProofOfTime:
+class RequestCompactProofOfTime(Streamable):
```
I would expect the annotation could add the base class too, no? (maybe for later)

rostislav replied:
The annotation could do the dataclass wrapping as well, but the comment in `def streamable` states that this is intended for linters so they could recognize the methods of the base class as well as the `__init__` method provided by `dataclass`.

arvidn commented on chia/wallet/wallet_puzzle_store.py:
```diff
@@ -142,11 +142,11 @@ async def get_derivation_record_for_puzzle_hash(self, puzzle_hash: str) -> Optio
-                row[0],
-                bytes.fromhex(row[2]),
+                uint32(row[0]),
+                bytes32.fromhex(row[2]),
```
we really store binary data in hex form in the database?

rostislav replied:
Yes, we currently do. See `add_derivation_paths()` in the same file.

arvidn commented on chia/util/streamable.py:
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
-            return bytes.decode(str_read_bytes, "utf-8")
-        raise RuntimeError(f"Type {f_type} does not have parse")
+            return parse_str
+        raise NotImplementedError(f"Type {f_type} does not have parse")
```
this message doesn't look right to me. Shouldn't it say "Type {f_type} does not have from_bytes()"?

arvidn replied:
I see, `parse()` is the main way of supporting the streamable protocol

rostislav replied:
Currently `from_bytes()` is used only for sized integer types. All other types that have custom serialization need to provide their own `.parse()` function.

arvidn reviewed (commented):
I think this approach is a lot nicer than the previous one. It especially makes it easy to test the parser for each primitive type, and really cover all the edge cases. I really think we should do that. This code has better work, everything else relies on it.

I would also have expected to see tests where we ensure that some (or maybe all) network messages encode to specific bytes. Just passing a round-trip test isn't enough to ensure we don't break the protocol.

I'll take a look at adding some tests today.

rostislav commented:
I performed the following benchmark on old and new Streamable to compare their performance: parsing the first 10000 mainnet blocks. The old result: 1.39s; the new result: 1.08s. The speed improvement is 23%.

arvidn committed:
additional unit tests for streamable parsers

mariano54 commented:
Nice work. If nothing is broken with the wallet, I think we can probable merge it (and if arvids tests pass)

arvidn committed:
add type annotations (#3222)

mariano54 reviewed (approved):
I've been running the spend bot with it, no issues
